File: source4/torture/rpc/dfs.c
/* 
   Unix SMB/CIFS implementation.
   test suite for lsa dfs operations

   Copyright (C) Andrew Tridgell 2003
   
   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.
   
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
   
   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#include "includes.h"


static BOOL test_Exist(struct dcerpc_pipe *p, TALLOC_CTX *mem_ctx)
{
	NTSTATUS status;
	struct dfs_Exist r;
	uint32 exist = 0;

	r.out.exist_flag = &exist;

	status = dcerpc_dfs_Exist(p, mem_ctx, &r);
	if (!NT_STATUS_IS_OK(status)) {
		printf("Exist failed - %s\n", nt_errstr(status));
		return False;
	}

	printf("exist=%d\n", exist);

	return True;
}

static BOOL test_Enum(struct dcerpc_pipe *p, TALLOC_CTX *mem_ctx)
{
	NTSTATUS status;
	struct dfs_Enum r;
	uint32 total=0;
	struct dfs_EnumStruct e;
	uint32 i = 0;
	struct dfs_String s;
	struct dfs_Enum1 e1;

	e.level = 1;
	e.e.enum1 = &e1;
	e.e.enum1->count = 0;
	e.e.enum1->s = &s;
	s.str = NULL;

	r.in.level = 1;
	r.in.bufsize = (uint32)-1;
	r.in.total = &total;
	r.in.unknown = NULL;
	r.in.info = &e;
	
	status = dcerpc_dfs_Enum(p, mem_ctx, &r);
	if (!NT_STATUS_IS_OK(status)) {
		printf("Enum failed - %s\n", nt_errstr(status));
		return False;
	}

	NDR_PRINT_DEBUG(dfs_EnumStruct, r.out.info);

	printf("total=%d\n", r.out.total?*r.out.total:-1);

	return True;
}

BOOL torture_rpc_dfs(int dummy)
{
        NTSTATUS status;
        struct dcerpc_pipe *p;
	TALLOC_CTX *mem_ctx;
	BOOL ret = True;

	mem_ctx = talloc_init("torture_rpc_dfs");

	status = torture_rpc_connection(&p, "netdfs");
	if (!NT_STATUS_IS_OK(status)) {
		return False;
	}
	
	if (!test_Exist(p, mem_ctx)) {
		ret = False;
	}

	if (!test_Enum(p, mem_ctx)) {
		ret = False;
	}

        torture_rpc_close(p);

	return ret;
}

File: python/samba/tests/dcerpc/dfs.py
#
# Unix SMB/CIFS implementation.
# Copyright Ralph Boehme <slow@samba.org> 2025
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

"""Tests for samba.dcerpc.dfs"""

import os
import logging
import samba
from samba.dcerpc import dfs
from samba.tests import RpcInterfaceTestCase
from samba.logger import get_samba_logger
from samba.credentials import Credentials
from samba.samba3 import libsmb_samba_internal as libsmb
import samba.tests.libsmb
from samba.samba3 import param as s3param

logger = get_samba_logger(name=__name__)

class DfsTests(samba.tests.libsmb.LibsmbTests):
    def setUp(self):
        super().setUp()
        self.dfs = dfs.netdfs('ncacn_np:%s[/pipe/netdfs]' % self.server, self.lp, self.creds)
        self.c = libsmb.Conn(self.server_ip, "msdfs-share", self.lp, self.creds)

    def tearDown(self):
        super().tearDown()

    def test_dfs_reparse_tag(self):
        self.dfs.Add('\\\\%s\\msdfs-share\\dfslink' % self.server, self.server, 'tmp', 'comment', 0)
        l = self.c.list('', info_level=libsmb.SMB2_FIND_ID_BOTH_DIRECTORY_INFO)
        files = {i['name']: i for i in l}
        self.assertEqual(files['dfslink']['reparse_tag'], libsmb.IO_REPARSE_TAG_DFS)
        self.dfs.Remove('\\\\%s\\msdfs-share\\dfslink' % self.server, self.server, 'tmp')
